[DNM] tracing: use LTTng for logging

mogeb opened the pull request:
See ``src/tracing/tracetool/README`` for details. Use ``-DWITH_LTTNG_LOGGING=ON`` with cmake to use LTTng for logging.
Using fio's objectstore backend, with the following job:
```
runtime=120
rw=randwrite
bs=4m
numjobs=16
iodepth=16
```
I get the following results:
- with dout and debug bluestore = 10/0:
```
Average throughput = 119 MiB/s
Average IOPS = 13.68
Disk usage: 48M
```

- with LTTng:
```
Average throughput = 125 MiB/s
Average IOPS = 14.09
Disk usage: 23M
```

fio is writing to Bluestore on an SSD and the log is being written to an HDD for both dout and LTTng. The throughput difference varies. I've seen LTTng up to 10% faster with more iodepth and numjobs (more contention on dout's internal structures), and I've seen it pretty much the same as dout (with lower iodepth and lower numjobs). However, the size on disk of the log is consistently 50%-60% smaller with LTTng (as expected).

liewegas commented on src/ceph_osd.cc:
```diff
@@ -615,6 +617,83 @@ int main(int argc, const char **argv)
+  chan.attr.read_timer_interval = 200;
+  chan.attr.output = LTTNG_EVENT_MMAP;
+
+  dout(1) << "mogeb creating session " << tracing_session << " in " << (string("/var/data/ceph/build/out/") + tracing_session) << dendl;
```
maybe replace "mogeb" with something else? :)

chardan commented on src/common/hobject.h:
```diff
@@ -445,6 +445,18 @@ struct ghobject_t {
 
+  operator std::string() {
+    std::stringstream out;
+    out << this;
```
It looks like this will write the address of "this" to the stream, rather than the value, *this. Is that what you want? Is there some other magic happening here that writes the value instead? Do I just need coffee? :-)

mogeb replied:
That is correct, it's for the implicit cast to string for unknown types. A simple `str()` method would also work, and we'd append the `.str()` call in tracetool.

chardan commented on src/osd/osd_types.h:
```diff
@@ -3792,6 +3809,7 @@ class pg_missing_const_i {
   virtual bool is_missing(const hobject_t& oid, pg_missing_item *out = nullptr) const = 0;
   virtual bool is_missing(const hobject_t& oid, eversion_t v) const = 0;
   virtual ~pg_missing_const_i() {}
+  virtual operator std::string() { return "needs implementation"; }
```
Given that this class is already pure virtual, why not just make this pure virtual as well?

mogeb replied:
You're obviously right. I was either not thinking, or this was a note to myself to look into implementing the operator in the base class. Either way, thanks!

tchaikov commented:
hey @mogeb may i learn your plan on this PR? please let me know if you need more hands. probably i can help on it.

mogeb commented:
@tchaikov sorry for the delay, I got side-tracked on another project for a couple of weeks now, but I'm planning on coming back to this PR next week.

I already have a second iteration of this PR ready and working, I just need to clean it a bit and probably ask you for a code review and to try it for usability. The second iteration gets rid of the `tracing/tracetool/subsys` files, as I explained [here](https://marc.info/?l=ceph-devel&m=152883737211746&w=2). If you have any thought that approach, please feel free to share it.

mogeb commented:
@tchaikov I've updated the PR. This is still experimental so please excuse some less than optimal code. You should be able to try it, and I think it's better to look at the usability before doing a full code review. For now, only `Bluestore.cc` can be instrumented, I've updated the `README` accordingly.

mogeb commented on src/common/hobject.cc:
```diff
@@ -355,7 +355,11 @@ int cmp(const hobject_t& l, const hobject_t& r)
 }
 
-
+ghobject_t::operator std::string() const {
```
Yes that would make sense. OTOH [this comment](https://github.com/ceph/ceph/pull/22458#discussion_r193863070) suggests to use a `str()` method instead of relying on the implicit `operator std::string()` to avoid surprises. What are your thoughts?

tchaikov replied:
i am also for the str() `method`, or at least make the `operator std::string` explicit, like 
```c++
explicit operator std::string() const { 
 /// ....
}
```

. @mogeb have you considered reusing a single trace point everywhere and send the fmt string to it, i.e. use it as an alternative of `syslog(3)` without the `...` parameters. if we are fine with this, we can avoid all these troubles. but the downside of this apporach.

mogeb replied:
@tchaikov sure, that's the very first solution I explored (pre-Cephalocon). I initially changed the implementation of `dout()` to just record a tracepoint. With that approach, none of the actual has to be changed. Another alternative could be to use LTTng's [`tracef()`](https://lttng.org/man/3/tracef/v2.8/) (similar to `printf`), but it also needs/uses strings.

My objective here though is to avoid passing strings around for performance reasons.

liewegas replied:
Right.  I think if the legacy dout lines degenerate into a slow string op, we can focus our efforts on replacing dout calls in the fast path and leave the less frequently-run and slow-path code alone until later (e.g., peering).

tchaikov replied:
thanks guys. after re-reading `tracing/tracetool/{README.md,parsetool.py`, now i understand the objective and rationale behind this design.

tchaikov replied:
@mogeb i agree with @liewegas  on this. and i think we can have a compromise here by differentiating following two use cases:

1. the logging in critical path: to use `trace_${event_name}(loglevel, subsys, [arg_type, arg_name]...)`, so we don't need to render the values of primitive types in string before passing them to LTTng.
1. the logging in non-critical path: to use `trace(loglevel, subsys, fmt_str, [arg]...)`. we can use `fmt::format()` offered by [libfmt](https://github.com/fmtlib/fmt) in the `pasrsetool.py`. as libfmt is already included by ceph as a submodule. and make `trace()` a template function. which calls `tracepoint(subsys, ${default_event_name}_at_${log_evel})` , assuming all combinations of `${default_event_name}_at_${log_level}` `TRACEPOINT_EVENT`s of different `${subsys}`s and `${log_level}`s combinations have been defined by `parsetool.py`. they only have a single `TP_FIELD`.

in this way, we can have

- a unified logging backend, so developers or users don't need to check different sources for tracing an issue, and
- a flexible front-end, so developers don't need to thinking about the name of a trace event, and the argument type of event payload. they only need to do so when it's needed.

what do you think?

mogeb replied:
I think the clang approach would be great. Sounds good, let's rethink it later as you said.

For having a critical and non-critical paths: just to make sure I understand, the non-critical path uses a formatted string using the arguments' ``operator <<()`` functions, correct? If that's the case, we could change the implementation of ``dout`` to record a tracepoint instead of using ``trace(loglevel, subsys, fmt_str, [arg]...)``. This becomes even easier for the developers as nothing changes: ``dout()`` just takes the string passed to it and ends up calling ``tracepoint()`` at the end. I'd need to figure out the details, but I was able to achieve that at the very beginning. Any thought?

tchaikov replied:
> the non-critical path uses a formatted string using the arguments' ``operator <<()`` functions, correct?

yes. at this moment, we can use the argument's `operator<<()`, so minimum change is needed. in short and long term, i'd suggest use libfmt's `fmt::format()` instead, it has better performance than that of `std::iostream`. probably we can expose two interfaces

- for the backward compatibility: `dout(1) << "ceph rock" << 5 << '!' << dendl;
- for the faster one: `trace(bluestore, 1, "{0} rock{1}{2}", "ceph", 5, '!');`
- for the fatest one: `trace_ceph_rocks(1, bluestore, string, who, "ceph", int, s, 5, uint_8, wow, '!');`

so, i'd suggest we should start using the `trace()` or `trace_${event_name}` interface in the new code. and replace the implementaion of `dout(1)` for the existing code.

tchaikov commented on src/tracing/tracetool/parsetool.py:
```diff
@@ -0,0 +1,472 @@
+                'static inline void __%(api)s(%(args)s)',
+                '{',
+                api=e.api(e.CEPH_TRACE),
+                args=", ".join([' '.join(arg[:2]) if is_valid_type(arg[0]) else ' '.join(('string', arg[1])) for arg in e.args]))
```
might want to differentiate the args which are already `std::string` or `string` in `Event.build()`. probably we could have three `Arg` types to represent them?

so we don't create a temporary copy of them just for passing them to `ctf_string()`.

tchaikov commented:
@mogeb i will try your patch out. probably the best way to improve it is to use it.

tchaikov commented on src/osd/CMakeLists.txt:
```diff
@@ -45,8 +45,15 @@ add_library(osd STATIC ${osd_srcs}
+if(WITH_LTTNG_LOGGING)
+  add_dependencies(osd bluestore-parsetool)
+  target_link_libraries(osd logging_tp)
+  target_link_libraries(osd lttng-ust-fork)
```
might want to add dependencies from `os` to `*_impl.h`, as `BlueStore.cc` now includes them. could use `set_source_files_properties()` for this purpose.

mogeb replied:
Noted, thanks.

mogeb commented:
@tchaikov any thought on the usability part?

tchaikov commented:
@mogeb i am stilling playing around with it. will reply you tomorrow.

tchaikov commented:
@mogeb i am following the instructions in `README.md`, what we have in the output of `lttng view` is:

```
...
[18:07:11.985680268] (+0.000163564) gen8 bluestore:do_write_small: { cpu_id = 4 }, { offset = 0x0, length = 0x1DC }
[18:07:11.985854796] (+0.000174528) gen8 bluestore:write: { cpu_id = 4 }, { cid = "meta", oid = "#-1:c683972f:::inc_osdmap.12:0#", offset = 0x0, length = 0x1D
C, r = 0 }
...
```

i think this the best compact representation of a (log) event we can have in the fast path, but from a typical user's point of view, it's still not very friendly. we can ship the `.tp` files in a separated package accompanied by a set of command line tools for analyzing the tracings. one of these tools will be able to act as a filter to utilize the babeltrace parser to compose the tracing events and the `.tp` file to produce a more readable output, in other words, to `printf` the fields along with the format string.

also, to integrate with existing logging infrastructure:

1. to see if we can use `lttng rotate` along with logrotate
1. update the systemd script to start/stop lttng sessions and tracing

but we can leave these tooling for following-up PRs. i guess, the first thing we need to have is `dout` backend and the "faster" mode, and to fix the `CMakeLists.txt`, to make sure it compiles if `WITH_LTTNG_LOGGING` is enabled. we can keep `WITH_LTTNG_LOGGING` disabled by default before we get these tooling and script sorted out.

once this change gets merged. we will

- improve the tooling, 
- update the systemd scripts.
- fix the packaging dependencies

and then switch OSD over to LTTng based logging.

what do you think?

mogeb commented:
@tchaikov I really like the approach, especially having a tool that knows how to rewrite the trace's events as more human-readable ones. Note however that the `.tp` files don't have the `printf` information in them (how to format and what to print), so we'll need to add a file for it. The session rotation option in the 2.10 version of LTTng certainly makes things easier. We can set the rotation to be automatic and either time-based or size-based. Did you measure any performance difference other than what I had reported? I think the reduced size on disk is a given, but let's confirm that the performance is also better to warrant this added complexity. Once I clean this PR I will run benchmarks on a real cluster too.

A couple of small things to clear:
- For the slow path, do we want to keep `dout` as it is (plain text logging), or do we want to make `dout` use LTTng, but with the complete formatted string in the payload of the event?
- Which log entries do we rewrite as fast-path? Maybe the ones lower than a certain loglevel?

tchaikov commented:
> Did you measure any performance difference other than what I had reported?

@mogeb not yet. i will try to reproduce your test in my test bed.

> For the slow path, do we want to keep dout as it is (plain text logging), or do we want to make dout use LTTng, but with the complete formatted string in the payload of the event?

i think it'd be hard for us to read the logging messages if we have a hybrid logging: plain text logging + LTTng tracing. to ease the transition. i'd recommend a compile-time option or runtime option for switching between the plain-text backend and the LTTng backend.

> Which log entries do we rewrite as fast-path? Maybe the ones lower than a certain loglevel?

i don't think there is a simple criteria for telling if a log entry is in the fast path by looking at its loglevel. for instance, we use `dout(1)` for a critical warning or very important message, for example, 
```c++
dout(1) << "not feeling good.. i am about to re-boot" << dendl;
```
and use `dout(30)` for a trivial message, 
```c++
dout(30) << "tick, tick. i am healthy!" << dendl;
```

i think, to ease the pain of migration, before we have the tool to "rewrite the trace's events as more human-readable ones.", we'd better leave the logging messages as they are. and leave the tool and the changes (for instance, ff26e95, 2b00fe2, 8e227fb) to LTTngize logging message for another PR.

tchaikov commented:
@mogeb in my case, lttng based log outperforms the existing dout(). tested with `test/fio/ceph-bluestore.fio`. a separated SSD for osd_data. and a separated HDD for log/ust.


```
lttng enable-event  -u -a

  WRITE: bw=65.0MiB/s (68.2MB/s), 65.0MiB/s-65.0MiB/s (68.2MB/s-68.2MB/s), io=1302MiB (1365MB), run=20018-20018msec
```

while with existing dout backend
```
  WRITE: bw=36.2MiB/s (37.0MB/s), 36.2MiB/s-36.2MiB/s (37.0MB/s-37.0MB/s), io=725MiB (760MB), run=20017-20017msec
```

mogeb commented:
@tchaikov so the performance is twice as good with LTTng compared to current dout? Is that using an actual OSD or fio with Bluestore backend? What's the loglevel you set for dout? In my tests, the logs were going to a different disk (which is more realistic) than the data with fio. I assume dout does more writes to disk than LTTng, so there's more contention. If the logs are going to the same disk than the data, the performance difference will be more flagrant. Nonetheless, this is good news, thanks.

Something came up on my side for so I got sidetracked for a few days, but I'll update this PR afterwards.

tchaikov commented:
> so the performance is twice as good with LTTng compared to current dout?

looks like.

> Is that using an actual OSD or fio with Bluestore backend?

fio w/ bluestore backend.


> What's the loglevel you set for dout? 

debug-bluestore = 10/0

> In my tests, the logs were going to a different disk (which is more realistic) than the data with fio. I assume dout does more writes to disk than LTTng, so there's more contention. If the logs are going to the same disk than the data, the performance difference will be more flagrant. Nonetheless, this is good news, thanks.

will double check my settings. and rerun the test .

> Something came up on my side for so I got sidetracked for a few days, but I'll update this PR afterwards.

sure. thanks!

cbodley commented on src/test/fio/fio_ceph_objectstore.cc:
```diff
@@ -20,6 +20,10 @@
+#ifndef typeof
+#define typeof __typeof__
+#endif
+
```
not sure why this is necessary. src/test/fio/CMakeLists.txt has a workaround for this:
```cmake
# fio headers use typeof(), which requires c++11 extensions
set_target_properties(fio_ceph_objectstore PROPERTIES
  CXX_EXTENSIONS ON
  COMPILE_FLAGS "${FIO_CFLAGS}")
```

mogeb replied:
This probably won't be it in the final version of this PR. However, I was planning on submitting it alone as the fio_ceph_objectstore.cc wouldn't build in my environment. I haven't tried it since though.

tchaikov replied:
i also ran into this issue while compiling `libfio_ceph_objectstore.so` with GCC 8.2, hence https://github.com/axboe/fio/pull/628 , which renders both workaround not necessary anymore.

tchaikov commented:
lttng
-----

  bw                           |                                         | io                | run                                       
------------------------------ | --------------------------------------- | ----------------- | -------------------
WRITE: bw=34.7MiB/s (36.4MB/s) | 34.7MiB/s-34.7MiB/s (36.4MB/s-36.4MB/s) | io=695MiB (729MB) | run=20018-20018msec
WRITE: bw=35.0MiB/s (36.7MB/s) | 35.0MiB/s-35.0MiB/s (36.7MB/s-36.7MB/s) | io=701MiB (735MB) | run=20016-20016msec
WRITE: bw=34.4MiB/s (36.1MB/s) | 34.4MiB/s-34.4MiB/s (36.1MB/s-36.1MB/s) | io=689MiB (723MB) | run=20016-20016msec
WRITE: bw=34.0MiB/s (35.7MB/s) | 34.0MiB/s-34.0MiB/s (35.7MB/s-35.7MB/s) | io=681MiB (714MB) | run=20014-20014msec
WRITE: bw=34.9MiB/s (36.6MB/s) | 34.9MiB/s-34.9MiB/s (36.6MB/s-36.6MB/s) | io=698MiB (732MB) | run=20014-20014msec

dout
----

  bw                           |                                         | io                | run                                       
------------------------------ | --------------------------------------- | ----------------- | -------------------
WRITE: bw=33.3MiB/s (34.9MB/s) | 33.3MiB/s-33.3MiB/s (34.9MB/s-34.9MB/s) | io=666MiB (698MB) | run=20011-20011msec
WRITE: bw=34.3MiB/s (35.9MB/s) | 34.3MiB/s-34.3MiB/s (35.9MB/s-35.9MB/s) | io=686MiB (720MB) | run=20032-20032msec
WRITE: bw=33.9MiB/s (35.6MB/s) | 33.9MiB/s-33.9MiB/s (35.6MB/s-35.6MB/s) | io=679MiB (712MB) | run=20019-20019msec
WRITE: bw=34.4MiB/s (36.1MB/s) | 34.4MiB/s-34.4MiB/s (36.1MB/s-36.1MB/s) | io=689MiB (723MB) | run=20011-20011msec
WRITE: bw=35.0MiB/s (37.7MB/s) | 35.0MiB/s-35.0MiB/s (37.7MB/s-37.7MB/s) | io=720MiB (755MB) | run=20023-20023msec

tchaikov commented:
@mogeb i reran the test. the result is not as good as last time. the performance lttng backend is slightly better than that of existing dout.

tchaikov commented:
what i don't understand is why my SSD is 2x slower than it was last week...  it was 
```bw=65.0MiB/s``` with LTTng logging, while it's now ```bw=34.7MiB/s ``` with the same settings.

tchaikov commented:
@mogeb ping?

mogeb commented:
@tchaikov there are now three ways to write an event:

1. Fast path: `trace_<event_name>(loglevel, subsys, type1, arg1, value1 [, type2, arg2, value2, ...], <fmt>);`
2. Slow path: `trace(loglevel, subsys, <libfmt-formatted string>, arg1, ...)`
3. Slow path: `trace_error(subsys, <libfmt-formatted string>, arg1, ...)`

3 is exactly like 2, but the loglevel is not required. For 1, there is no string manipulation, and raw basic types have to be provided. The result in the trace is a compact event. For 2 and 3, arguments are formatted to a string and passed to the event.

The only issue right now is that for 2 and 3, you have to use `fmt::ptr(arg)` if `arg` is a pointer, as explained [here](https://github.com/fmtlib/fmt/issues/415#issuecomment-328350139). It doesn't look like there's a straight way to print pointers (at least in C++).

mogeb commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -7365,12 +7450,18 @@ int BlueStore::read(
-  dout(10) << __func__ << " " << cid << " " << oid
-	   << " 0x" << std::hex << offset << "~" << length << std::dec
-	   << " = " << r << dendl;
+  trace(10, "bluestore", "read {}, {}, {}, {} and {}",
```
We could also replace `susbsys` with `__func__`.

mogeb commented:
Rebased. I'm trying to run this on a real cluster now.

tchaikov commented:
sorry for the latency, i will review this PR early tomorrow.

mogeb commented:
Update: my hardware is currently broken so I'm waiting for it to be fixed to report performance numbers on a real cluster. I'll report soon and sorry for the delay.

mogeb commented:
New results [here](https://docs.google.com/spreadsheets/d/1NfNBV8sm1_tgdBYmMAfmoEufZkkUNpkJNZ0gsOK4RgE/edit?usp=sharing)

jdurgin commented:
This was discussed at the perf weekly 2019-06-20, where @mogeb shared those results. Given the difficulty of use of lttng for both users and developers, we decided that it would make more sense to concentrate on optimizing the log statements in the hot path (e.g. reducing the memcpys and stringification of many structures there) rather than converting the logging machinery to lttng.

I'd update the trello card earlier: https://trello.com/c/WAu7nsH4/362-osd-make-logging-on-the-i-o-path-lightweight but this PR being open caused some confusion, so closing it now.

cbodley commented:
@mogeb thanks for all of your efforts in exploring this!

mogeb commented:
@cbodley and @jdurgin thanks for your help as well.
